### ISSUE TITLE
EmailAddressAttribute behaves differently in .NET Core vs. NetFx

### ISSUE THREAD

**Issue by Tornhoof** [dotnet/corefx#28501]
It behaves differently for the following Unit Tests and I think both behaviours are not 100% correct:
Just for reference:
https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations.emailaddressattribute(v=vs.110).aspx

```csharp
public class UnitTest1
{
    private static EmailAddressAttribute AddressAttribute = new EmailAddressAttribute();

    [Theory]
    [InlineData("test@test.com", true)]
    [InlineData("test@test.co.uk", true)]
    [InlineData("test+t@test.co.uk", true)]
    [InlineData("test", false)]
    [InlineData("test.com", false)]
    [InlineData("test@test", false)]
    [InlineData("test@test.", false)]
    public void EmailValidity(string email, bool result)
    {
        Assert.Equal(result, AddressAttribute.IsValid(email));
    }
}
```
In NetFX 4.7 all tests succeed.
In NetCore 1.0/2.0 the tests for "test@test" and "test@test." fail.

In #25129 @PeterSmithRedmond said that atleast test@test is a valid email, but as far as i understand the spec test@test.  (notice the dot) is not, the domain part is called dot-atom in the spec https://tools.ietf.org/html/rfc2822#section-3.2.4 and has either no dot or dot and other valid symbols after the dot.

If I'm correct there are bugs in both implementations, in netfx the test for "test@test" should fail (because it's a valid address) and in net core the test for "test@test." should succeed, because it's not a valid address.

Repro:

[EmailAddressAttributeRepro.zip](https://github.com/dotnet/corefx/files/1850894/EmailAddressAttributeRepro.zip)

Just run it with the different test frameworks.

**Comment by Tornhoof** [comment]
If this is again a duplicate of https://github.com/dotnet/corefx/issues/27323, then please add a doc comment  regarding the checking algorithm, the comment how it works is only an inline comment somwhere in the method.

**Comment by wfurt** [comment]
The '.' at the end means root of the DNS tree. 'test.' is top level domain (like 'com.') where as 
'test' is unqualified host test. So the meaning is different even  if the string looks similar.

**Comment by PeterSmithRedmond** [comment]
Looking again at the email addresses: AFAICT, wfurt is correct. Going back to some of the earlier RFCS, [RFC 822](https://tools.ietf.org/html/rfc822) is pretty clear that you can't end a domain name with a period. The email-oriented RFCs agree: [RFC 5322](https://tools.ietf.org/html/rfc5322) section 3.4.1, Addr-spec specification concurs: you can't end a domain name with a dot.

Just to make life confusing, per the wiki article on FQDN, some addresses have periods at the end. But per [RFC 1123](https://tools.ietf.org/html/rfc1123#page-13) section 5.2.18, 
 
> Some systems over-qualify domain names by adding a
>  trailing dot to some or all domain names in addresses or
>  message-ids.  This violates RFC-822 syntax.
> 

TL/DR: I think that test@test. should not be considered a valid email address. Almost all of the RFCs say it's wrong. Part of the value of the validate test to at least partially validate the user's input; haivn g adot at the end is almost certainly an unwanted typo and not some super-clever RFC hack.

**Comment by Tornhoof** [comment]
I can provide a PR for the fix in the EmailAddressAttribute.cs, the corefx version does not have the monstrous regex, so it should be fairly simple to check that the last char is not a '.'.
I would add two test cases:
1. Failure Case: Testing that "test@test." is not valid
2. Success Case: Testing that "test@test" is valid (this one is missing from the tests in https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Annotations/tests/EmailAddressAttributeTests.cs, the success test cases all have '.' somewhere in there)

**Comment by ajcvickers** [comment]
@Tornhoof That sounds like a reasonable change, however since we are closing down on 2.1 features I'm going to move this to Future for now. @divega @lajones Any thoughts?

**Comment by divega** [comment]
@ajcvickers I agree. No other thoughts.

**Comment by lajones** [comment]
@ajcvickers @divega Agreed. Nothing more to add.

**Comment by wfurt** [comment]
note that I tried Outlook and unix "mail" client and both takes address with trailing '.'

**Comment by Tornhoof** [comment]
K9-Mail on Android accepts it but strips the dot before sending.
Alpine accepts "test@test.local." and rewrites it to "test@test.local..test.local" and apparently postfix can handle that.

**Comment by lukos** [comment]
Can someone point to the storyline for this? As far as I can see, MS made a massive breaking change from a reasonable regex-based attribute (not perfect for sure) to an attribute that literally checks that there is exactly one @ symbol, which is not at the start or the end, which make the whole thing useless and has broken a raft of unit tests and will also provide many false positives. Fortunately we spotted this when porting a project but many people could easily miss this and it would introduce many bugs as a result. A property could be used to flag whether to use the regex or a more simple check?

**Comment by ajcvickers** [comment]
@lukos The discussion is in #17873

**Comment by wfurt** [comment]
also specific examples would be helpful @lukos

**Comment by lukos** [comment]
Two examples:

.something@example.com
something.@example.com

I believe both should fail validation and neither do. These are also reasonable typing errors to make.

**Comment by wfurt** [comment]
rfc3696, section 3. 
And underlying rfc2821

```
Mailbox = Local-part "@" Domain
      Local-part = Dot-string / Quoted-string
            ; MAY be case-sensitive
      Dot-string = Atom *("." Atom)
      Atom = 1*atext
      Quoted-string = DQUOTE *qcontent DQUOTE
      String = Atom / Quoted-string

```

as @lukos mentioned,  spaces should be legal if quoted.

**Comment by ajcvickers** [comment]
Closing this as we don't plan to make any changes to the email validation.  See https://github.com/dotnet/corefx/issues/32740#issuecomment-466578595
